Cannot compile classes deriving from ObservableValidator with generic type parameter

### Describe the bug

Compilation fails if you create a class that dervies from ObservableValidator and has a generic type parameter.

Output includes:

```
1>C:\dev\ObservableValidatorGeneric\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.ObservableValidatorValidateAllPropertiesGenerator\.Class1-1.cs(9,97,9,98): error CS0246: The type or namespace name 'T' could not be found (are you missing a using directive or an assembly reference?)
```

### Regression

_No response_

### Steps to reproduce

```text
using CommunityToolkit.Mvvm.ComponentModel;
namespace ObservableValidatorGeneric;
public class Class1<T> : ObservableValidator { }
```
```


### Expected behavior

Compiles

### Screenshots

_No response_

### IDE and version

VS 2022

### IDE version

17.1.6

### Nuget packages

- [ ] CommunityToolkit.Common
- [ ] CommunityToolkit.Diagnostics
- [ ] CommunityToolkit.HighPerformance
- [X] CommunityToolkit.Mvvm (aka MVVM Toolkit)

### Nuget package version(s)

8.0.0-preview3

### Additional context

Presumably, the generator needs to emit a generic type parameter here? --

```
    partial class __ObservableValidatorExtensions
    {
        [global::System.ComponentModel.EditorBrowsable(global::System.ComponentModel.EditorBrowsableState.Never)]
        [global::System.Obsolete("This method is not intended to be called directly by user code")]
        public static global::System.Action<object> CreateAllPropertiesValidator(global::ObservableValidatorGeneric.Class1<T> _)
        {
            static void ValidateAllProperties(object obj)
            {
                var instance = (global::ObservableValidatorGeneric.Class1<T>)obj;
            }

            return ValidateAllProperties;
        }
    }
```


### Help us help you

No, just wanted to report this

--- a/CommunityToolkit.Mvvm.SourceGenerators/ComponentModel/ObservableValidatorValidateAllPropertiesGenerator.cs
+++ b/CommunityToolkit.Mvvm.SourceGenerators/ComponentModel/ObservableValidatorValidateAllPropertiesGenerator.cs
@@ -28,7 +28,7 @@ public sealed partial class ObservableValidatorValidateAllPropertiesGenerator : 
             .CreateSyntaxProvider(
                 static (node, _) => node is ClassDeclarationSyntax,
                 static (context, _) => (context.Node, Symbol: (INamedTypeSymbol)context.SemanticModel.GetDeclaredSymbol(context.Node)!))
-            .Where(static item => !item.Symbol.IsAbstract && item.Node.IsFirstSyntaxDeclarationForSymbol(item.Symbol))
+            .Where(static item => item.Symbol is { IsAbstract: false, IsGenericType: false } && item.Node.IsFirstSyntaxDeclarationForSymbol(item.Symbol))
             .Select(static (item, _) => item.Symbol);
 
         // Get the types that inherit from ObservableValidator and gather their info

--- a/CommunityToolkit.Mvvm.SourceGenerators/Messaging/IMessengerRegisterAllGenerator.cs
+++ b/CommunityToolkit.Mvvm.SourceGenerators/Messaging/IMessengerRegisterAllGenerator.cs
@@ -31,7 +31,7 @@ public sealed partial class IMessengerRegisterAllGenerator : IIncrementalGenerat
             .CreateSyntaxProvider(
                 static (node, _) => node is ClassDeclarationSyntax,
                 static (context, _) => (context.Node, Symbol: (INamedTypeSymbol)context.SemanticModel.GetDeclaredSymbol(context.Node)!))
-            .Where(static item => !item.Symbol.IsAbstract && item.Node.IsFirstSyntaxDeclarationForSymbol(item.Symbol))
+            .Where(static item => item.Symbol is { IsAbstract: false, IsGenericType: false } && item.Node.IsFirstSyntaxDeclarationForSymbol(item.Symbol))
             .Select(static (item, _) => item.Symbol);
 
         // Get the target IRecipient<TMessage> interfaces and filter out other types

--- a/CommunityToolkit.Mvvm/ComponentModel/ObservableValidator.cs
+++ b/CommunityToolkit.Mvvm/ComponentModel/ObservableValidator.cs
@@ -606,8 +606,8 @@ public abstract class ObservableValidator : ObservableObject, INotifyDataErrorIn
                 from property in validatableProperties
                 select Expression.Call(inst0, validateMethod, new Expression[]
                 {
-                        Expression.Convert(Expression.Call(inst0, property.GetMethod), typeof(object)),
-                        Expression.Constant(property.Name)
+                    Expression.Convert(Expression.Call(inst0, property.GetMethod), typeof(object)),
+                    Expression.Constant(property.Name)
                 }));
 
             return Expression.Lambda<Action<object>>(body, arg0).Compile();

--- a/tests/CommunityToolkit.Mvvm.UnitTests/Test_Messenger.cs
+++ b/tests/CommunityToolkit.Mvvm.UnitTests/Test_Messenger.cs
@@ -3,11 +3,13 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Collections.Generic;
 using System.Linq;
 using System.Reflection;
 using System.Threading;
 using System.Threading.Tasks;
 using CommunityToolkit.Mvvm.Messaging;
+using CommunityToolkit.Mvvm.Messaging.Messages;
 using Microsoft.VisualStudio.TestTools.UnitTesting;
 
 namespace CommunityToolkit.Mvvm.UnitTests;
@@ -370,6 +372,40 @@ public partial class Test_Messenger
 
         Assert.IsFalse(messenger.IsRegistered<MessageA>(recipient));
         Assert.IsFalse(messenger.IsRegistered<MessageB>(recipient));
+    }
+
+    // See https://github.com/CommunityToolkit/dotnet/issues/198
+    [TestMethod]
+    [DataRow(typeof(StrongReferenceMessenger))]
+    [DataRow(typeof(WeakReferenceMessenger))]
+    public void Test_Messenger_IRecipient_SomeMessages_WithGenericTypeParameters(Type type)
+    {
+        IMessenger? messenger = (IMessenger)Activator.CreateInstance(type)!;
+        GenericRecipient<string> recipient = new();
+
+        messenger.RegisterAll(recipient);
+
+        Assert.IsTrue(messenger.IsRegistered<GenericRecipient<string>.Message>(recipient));
+        Assert.IsTrue(messenger.IsRegistered<ValueChangedMessage<string>>(recipient));
+        Assert.IsTrue(messenger.IsRegistered<MessageA>(recipient));
+
+        Assert.IsNull(recipient.ReceivedMessage1);
+        Assert.IsNull(recipient.ReceivedMessage2);
+        Assert.IsNull(recipient.ReceivedMessage3);
+
+        GenericRecipient<string>.Message message1 = messenger.Send(new GenericRecipient<string>.Message(new List<string> { "Hello world!" }));
+        ValueChangedMessage<string> message2 = messenger.Send(new ValueChangedMessage<string>("Hello"));
+        MessageA message3 = messenger.Send<MessageA>();
+
+        Assert.AreSame(recipient.ReceivedMessage1, message1);
+        Assert.AreSame(recipient.ReceivedMessage2, message2);
+        Assert.AreSame(recipient.ReceivedMessage3, message3);
+
+        messenger.UnregisterAll(recipient);
+
+        Assert.IsFalse(messenger.IsRegistered<GenericRecipient<string>.Message>(recipient));
+        Assert.IsFalse(messenger.IsRegistered<ValueChangedMessage<string>>(recipient));
+        Assert.IsFalse(messenger.IsRegistered<MessageA>(recipient));
     }
 
     [TestMethod]
@@ -1058,5 +1094,37 @@ public partial class Test_Messenger
 
     public sealed class MessageB
     {
+    }
+
+    public partial class GenericRecipient<T> : IRecipient<GenericRecipient<T>.Message>, IRecipient<ValueChangedMessage<T>>, IRecipient<MessageA>
+    {
+        public Message? ReceivedMessage1 { get; private set; }
+
+        public ValueChangedMessage<T>? ReceivedMessage2 { get; private set; }
+
+        public MessageA? ReceivedMessage3 { get; private set; }
+
+        public void Receive(Message message)
+        {
+            ReceivedMessage1 = message;
+        }
+
+        public void Receive(ValueChangedMessage<T> message)
+        {
+            ReceivedMessage2 = message;
+        }
+
+        public void Receive(MessageA message)
+        {
+            ReceivedMessage3 = message;
+        }
+
+        public sealed class Message : ValueChangedMessage<List<T>>
+        {
+            public Message(List<T> list)
+                : base(list)
+            {
+            }
+        }
     }
 }

--- a/tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservableValidator.cs
+++ b/tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservableValidator.cs
@@ -539,6 +539,32 @@ public class Test_ObservableValidator
         Assert.IsNull(createAllPropertiesValidatorMethod);
     }
 
+    // See https://github.com/CommunityToolkit/dotnet/issues/246
+    [TestMethod]
+    public void Test_ObservableValidator_WithGenericTypeParameters()
+    {
+        GenericPerson<string> model = new();
+
+        model.Name = "Bob";
+
+        model.ValidateAllProperties();
+
+        Assert.IsTrue(model.HasErrors);
+
+        ValidationResult[] errors = model.GetErrors(nameof(model.Value)).ToArray();
+
+        Assert.IsNotNull(errors);
+        Assert.AreEqual(errors.Length, 1);
+
+        CollectionAssert.AreEqual(errors[0].MemberNames.ToArray(), new[] { nameof(model.Value) });
+
+        model.Value = "Ross";
+
+        model.ValidateAllProperties();
+
+        Assert.IsFalse(model.HasErrors);
+    }
+
     public class Person : ObservableValidator
     {
         private string? name;
@@ -796,5 +822,20 @@ public class Test_ObservableValidator
         [Required]
         [MinLength(2)]
         public string? Name { get; set; }
+    }
+
+    public class GenericPerson<T> : ObservableValidator
+    {
+        [Required]
+        [MinLength(1)]
+        public string? Name { get; set; }
+
+        [Required]
+        public T? Value { get; set; }
+
+        public new void ValidateAllProperties()
+        {
+            base.ValidateAllProperties();
+        }
     }
 }

